File: Day_7/threadsafe.java
/*
   Day 7: Thread-Safe ArrayList

   This is a class that acts as a normal
   array, but works with multithreading.
*/
import java.util.ArrayList;
import java.util.concurrent.locks.*;
import java.lang.Thread;

public class threadsafe
{
   public static class ThreadSafeArrayList<T>
   {
      // Default constructor
      public ThreadSafeArrayList()
      {
         arr = new ArrayList();
         lock = new ReentrantLock();
      }
      public ThreadSafeArrayList(T[] array)
      {
         arr = new ArrayList();
         lock = new ReentrantLock();

         // Find the size of the array, if it is greater than 0, initialize
         // the vector to the array.
         if(array.length!=0)
            for(int i = 0; i < array.length;i++)
               push_back(array[i]);
      }

      // These are read-only functions, so we do not need it to be synchronized
      public T at(Integer i){ return arr.get(i); }
      public Integer size(){ return arr.size(); }
      public Boolean empty(){ return arr.isEmpty(); }
      public T front(){ return at(0); }
      public T back(){ return at(size()-1); }

      // Only one thread can add at a time
      void push_back(T elem)
      {
         lock.lock();
         arr.add(elem);
         lock.unlock();
      }
      void pop_back()
      {
         lock.lock();
         arr.remove(size()-1);
         lock.unlock();
      }
      void insert_at(Integer i, T elem)
      {
         lock.lock();
         arr.add(i,elem);
         lock.unlock();
      }
      void erase_at(Integer i)
      {
         lock.lock();
         arr.remove(i);
         lock.unlock();
      }
      void clear()
      {
         lock.lock();
         arr.clear();
         lock.unlock();
      }

      private ArrayList<T> arr;
      private Lock lock;
   }

   public static class addThread implements Runnable
   {
      addThread(ThreadSafeArrayList<Integer> arr, Integer start, Integer end)
      {
         this.arr = arr;
         this.start = start;
         this.end = end;
      }
      public void run()
      {
         for(int i = start; i < end; i++)
            arr.push_back(i);
      }
      private ThreadSafeArrayList<Integer> arr;
      private Integer start;
      private Integer end;
   }

   public static void main(String[] args)
   {
      ThreadSafeArrayList<Integer> arr = new ThreadSafeArrayList<>();
      ArrayList<Thread >threads = new ArrayList();

      // Populate thread safe vector with elements
      for(int i = 0; i < 5; i++)
      {
         threads.add(i,new Thread(new addThread(arr,i*2*1000000,(i+1)*2*1000000)));
         threads.get(i).start();
      }
   }
}

File: Day_7/threadsafe.cpp
/*
   Day 7: Thread-Safe Vector

   This is a class that acts as a normal
   array, but works with multithreading.

   Compile with:
   g++ -std=c++11 -pthread threadsafe.cpp
*/
#include <iostream>
#include <mutex>
#include <vector>
#include <thread>
#include <functional>
using namespace std;

template <class T>
class ThreadSafeVector
{
   public:
      // Default constructor
      ThreadSafeVector(){}

      ThreadSafeVector(T array[])
      {
         // Find the size of the array, if it is greater than 0, initialize
         // the vector to the array.
         int n = sizeof(array) == 0 ? 0 : sizeof(array)/sizeof(array[0]);
         if(n!=0)
            for(int i = 0; i < n; i++)
               push_back(array[i]);
      }

      // These are read-only functions, so we do not need it to be synchronized
      T at(int i){ return vect[i]; }
      T &operator[](int i){ return vect[i]; }
      int size(){ return vect.size(); }
      int max_size(){ return vect.max_size(); }
      bool empty(){ return vect.empty(); }
      T front(){ return vect.front(); }
      T back(){ return vect.back(); }

      // Only one thread can add at a time
      void push_back(T elem)
      {
         lock.lock();
         vect.push_back(elem);
         lock.unlock();
      }
      void pop_back()
      {
         lock.lock();
         vect.pop_back();
         lock.unlock();
      }
      void insert_at(int i, T elem)
      {
         lock.lock();
         vect.insert(vect.begin()+i,elem);
         lock.unlock();
      }
      void erase_at(int i)
      {
         lock.lock();
         vect.erase(vect.begin()+i);
         lock.unlock();
      }
      void clear()
      {
         lock.lock();
         vect.clear();
         lock.unlock();
      }

   private:
      vector<T> vect;
      mutex lock;
};

void addObjectsInRange(ThreadSafeVector<int>& vect, int start, int end)
{
   for(int i = start; i < end; i++)
      vect.push_back(i);
}

int main()
{
   ThreadSafeVector<int> vect;
   vector <thread> threads;

   // Populate thread safe vector with elements
   for(int i = 0; i < 5; i++)
      threads.push_back(thread(addObjectsInRange,ref(vect),i*2*1000000,(i+1)*2*1000000));
   for(int i = 0; i < 5; i++)
      threads[i].join();
   return 0;
}
